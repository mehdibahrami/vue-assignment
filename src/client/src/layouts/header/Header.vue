<template>
  <header>
    <!-- Navbar -->
    <nav class="navbar">
      <div class="inner-navbar">
        <div class="container">
          <div class="main-menu">
            <!-- Logo -->
            <img class="logo" :src="logo" alt="ViriCiti" />
            <!-- Horizontal navbar menu -->
            <HorizontalNavbarMenu :items="navbarMenuItems" />
          </div>
          <!-- Mobile menu -->
          <div class="mobile-menu">
            <!-- Mobile menu button -->
            <button
              type="button"
              class="mobile-menu-button"
              @click="isMenuOpen = !isMenuOpen"
            >
              <img
                v-if="!isMenuOpen"
                :src="menuIcon"
                class="mobile-menu-icon"
              />
              <img v-else :src="closeIcon" class="mobile-menu-icon" />
            </button>
          </div>
        </div>
      </div>
      <!-- Vertical navbar menu -->
      <VerticalNavbarMenu :isOpen="isMenuOpen" :items="navbarMenuItems" />
    </nav>
  </header>
</template>

<script>
import { HorizontalNavbarMenu, VerticalNavbarMenu } from "@/layouts";
import navbarMenuItems from "@/layouts/header/navbarMenuItems.js";

export default {
  name: "Header",
  components: {
    HorizontalNavbarMenu,
    VerticalNavbarMenu,
  },
  data() {
    return {
      logo: require("@/assets/img/logo.svg"),
      menuIcon: require("@/assets/img/menu.svg"),
      closeIcon: require("@/assets/img/close.svg"),
      isMenuOpen: false,
      navbarMenuItems,
    };
  },
};
</script>

<style scoped lang="scss">
.navbar {
  background-color: $primary;
  padding-bottom: 8rem;
}

.inner-navbar {
  max-width: 80rem;
  margin-left: auto;
  margin-right: auto;
}

.container {
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 4rem;
  padding-left: 1rem;
  padding-right: 1rem;
  border-bottom-width: 1px;
  border-color: $green-dark;
}

.main-menu {
  display: flex;
  align-items: center;
}

.logo {
  height: 2rem;
}

.mobile-menu {
  margin-right: -0.5rem;
  display: flex;
}

.mobile-menu-button {
  background-color: $green-dark;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0.5rem;
  border-radius: 0.375rem;
  color: $white;
}

.mobile-menu-button:hover {
  background-color: $green-darker;
}

.mobile-menu-icon {
  display: block;
  height: 1.5rem;
  width: 1.5rem;
}

@media (min-width: 640px) {
  .inner-navbar {
    padding-left: 1.5rem;
    padding-right: 1.5rem;
  }

  .container {
    padding-left: 0px;
    padding-right: 0px;
  }
}

@media (min-width: 768px) {
  .mobile-menu {
    display: none;
  }
}

@media (min-width: 1024px) {
  .inner-navbar {
    padding-left: 2rem;
    padding-right: 2rem;
  }
}
</style>
