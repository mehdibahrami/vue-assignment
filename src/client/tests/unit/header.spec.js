import { shallowMount } from "@vue/test-utils";
import Header from "@/layouts/header/Header.vue";

describe("Header", () => {
  let wrapper;

  beforeEach(() => {
    // shallow mounting the main component
    wrapper = shallowMount(Header);
  });

  it("renders correctly", () => {
    expect(
      wrapper.findComponent({ name: "HorizontalNavbarMenu" }).exists()
    ).toBeTruthy();
    expect(
      wrapper.findComponent({ name: "VerticalNavbarMenu" }).exists()
    ).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("checks logo image exists", () => {
    const logo = wrapper.findAll("img.logo");
    expect(logo.length).toBe(1);
  });

  it("by default, the menu should be closed", () => {
    expect(wrapper.vm.isMenuOpen).toBe(false);
  });

  it("opens menu when pressing the menu button", () => {
    const menuButton = wrapper.find(".mobile-menu-button");
    menuButton.trigger("click");
    expect(wrapper.vm.isMenuOpen).toBe(true);
  });

  it("closes menu when the menu is open and pressing menu button", async () => {
    await wrapper.setData({ isMenuOpen: true });
    const menuButton = wrapper.find(".mobile-menu-button");
    menuButton.trigger("click");
    expect(wrapper.vm.isMenuOpen).toBe(false);
  });
});
