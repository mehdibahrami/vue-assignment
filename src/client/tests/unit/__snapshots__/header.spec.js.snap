// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<header>
  <!-- Navbar -->
  <nav
    class="navbar"
  >
    <div
      class="inner-navbar"
    >
      <div
        class="container"
      >
        <div
          class="main-menu"
        >
          <!-- Logo -->
          <img
            alt="ViriCiti"
            class="logo"
            src=""
          />
          <!-- Horizontal navbar menu -->
          <horizontal-navbar-menu-stub
            items="[object Object]"
          />
        </div>
        <!-- Mobile menu -->
        <div
          class="mobile-menu"
        >
          <!-- Mobile menu button -->
          <button
            class="mobile-menu-button"
            type="button"
          >
            <img
              class="mobile-menu-icon"
              src=""
            />
          </button>
        </div>
      </div>
    </div>
    <!-- Vertical navbar menu -->
    <vertical-navbar-menu-stub
      isopen="false"
      items="[object Object]"
    />
  </nav>
</header>
`;
